src/sentry/api/endpoints/project_artifact_bundle_files.py

```diff
@@ -0,0 +1,111 @@
+
+
+@region_silo_endpoint
+class ProjectArtifactBundleFilesEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectReleasePermission,)
+    rate_limits = RateLimitConfig(
+        group="CLI", limit_overrides={"GET": SENTRY_RATELIMITER_GROUP_DEFAULTS["default"]}
+    )
+
+    def get(self, request: Request, project, bundle_id) -> Response:
+        """
+        List a Project Artifact Bundle's Files
+        ``````````````````````````````
+
+        Retrieve a list of artifact bundle files for a given artifact bundle.
+
+        :pparam string organization_slug: the slug of the organization the
+                                          release belongs to.
+        :pparam string project_slug: the slug of the project to list the
+                                     release files of.
+        :pparam string bundle_id: bundle_id of the artifact bundle to read contents from.
+        """
+
+        try:
+            artifact_bundle = ArtifactBundle.objects.get(
+                organization_id=project.organization.id, bundle_id=bundle_id
+            )
+        except ArtifactBundle.DoesNotExist:
+            return Response(
+                {"error": f"The artifact bundle with {bundle_id} does not exist"}, status=404
+            )
+
+        try:
+            ProjectArtifactBundle.objects.get(
+                project_id=project.id, artifact_bundle=artifact_bundle
+            )
+        except ProjectArtifactBundle.DoesNotExist:
+            return Response(
+                {"error": f"The artifact bundle with {bundle_id} is not bound to this project"},
+                status=400,
+            )
+
+        try:
+            # We open the archive to fetch the number of files.
+            archive = ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=False)
+        except Exception:
+            return Response({"error": "The artifact bundle archive can't be opened"})
+
+        def expose_artifact_bundle_file(file_path, info):
+            headers = archive.normalize_headers(info.get("headers", {}))
+            debug_id = archive.normalize_debug_id(headers.get("debug-id"))
+            file_info = archive.get_file_info(file_path)
+
+            return {
+                "id": base64.b64encode(bytes(file_path.encode("utf-8"))),
+                "type": info.get("type"),
+                "filePath": file_path,
+                "debugId": debug_id,
+                "dateCreated": artifact_bundle.date_added.isoformat().replace("+00:00", "Z"),
+                "size": file_info.file_size if file_info is not None else None,
+            }
+
+        def serialize_results(r):
+            serialized_results = serialize(
+                [expose_artifact_bundle_file(file_path, info) for file_path, info in r],
+                request.user,
+            )
+            # We must close the archive once all the results have been fetched, otherwise we will get an error.
+            archive.close()
```

Comment:
Is it possible that this flow (or `expose_artifact_bundle_file` itself) will somehow fail along the way and archive will be kept opened? Can we do something like `finally` in python? Is it also possible that `paginate` will fail and we should handle this case and close archive then as well?

Reply:
I hate doing this `.close` here and was wondering the same thing. I might have to see what we can do in cases of failure. I might wrap the pagination with an error handling in place, or actually, close the archive before returning the resultset.